core/drivers/stm32_cpu_opp.c

```diff
@@ -0,0 +1,509 @@
+ * supported value and still return true.
+ */
+static bool opp_voltage_is_supported(struct regulator *regul, uint32_t *volt_uv)
+{
+	int target_volt_uv = 0;
+	int new_volt_uv = 0;
+	int min_uv = 0;
+	int max_uv = 0;
+	struct regulator_voltages_desc *desc = NULL;
+	const int *levels = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(*volt_uv < INT_MAX);
+	target_volt_uv = (int)*volt_uv;
+
+	res = regulator_supported_voltages(regul, &desc, &levels);
+	if (res) {
+		regulator_get_range(regul, &min_uv, &max_uv);
+		if (target_volt_uv > max_uv)
+			return false;
+		if (target_volt_uv < min_uv)
+			*volt_uv = min_uv;
+		return true;
+	}
+
+	if (desc->type == VOLTAGE_TYPE_FULL_LIST) {
+		unsigned int i = 0;
+
+		for (i = 0 ; i < desc->num_levels; i++) {
+			if (levels[i] >= target_volt_uv) {
+				new_volt_uv = levels[i];
+				break;
+			}
+		}
+		if (new_volt_uv == 0)
+			return false;
+
+	} else if (desc->type == VOLTAGE_TYPE_INCREMENT) {
+		int min = levels[0];
+		int max = levels[1];
+		int step = levels[2];
+
+		if (new_volt_uv > max)
+			return false;
+
+		new_volt_uv = min +
+			      DIV_ROUND_UP(target_volt_uv - min, step) * step;
+	} else {
+		return false;
+	}
+
+	*volt_uv = new_volt_uv;
+
+	return true;
+}
+
+static TEE_Result set_clock_then_voltage(unsigned int opp)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = set_opp_clk_rate(opp);
+	if (res)
+		return res;
+
+#ifdef CFG_STM32MP13
+	if (cpu_opp.dvfs[opp].volt_uv <= PWR_MPU_RAM_LOW_SPEED_THRESHOLD)
+		io_setbits32(stm32_pwr_base() + PWR_CR1_OFF,
+			     PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+	res = set_opp_voltage(opp);
+	if (res) {
+		/* Restore previous OPP */
+#ifdef CFG_STM32MP13
+		if (cpu_opp.dvfs[cpu_opp.current_opp].volt_uv >
+		    PWR_MPU_RAM_LOW_SPEED_THRESHOLD)
+			io_clrbits32(stm32_pwr_base() + PWR_CR1_OFF,
+				     PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+		if (set_opp_clk_rate(cpu_opp.current_opp))
+			panic();
+
+		return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result set_voltage_then_clock(unsigned int opp)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = set_opp_voltage(opp);
+	if (res)
+		return res;
+
+#ifdef CFG_STM32MP13
+	if (cpu_opp.dvfs[opp].volt_uv > PWR_MPU_RAM_LOW_SPEED_THRESHOLD)
+		io_clrbits32(stm32_pwr_base() + PWR_CR1_OFF,
+			     PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+	res = set_opp_clk_rate(opp);
+	if (res) {
+		/* Restore previous OPP */
+#ifdef CFG_STM32MP13
+		if (cpu_opp.dvfs[cpu_opp.current_opp].volt_uv <=
+		    PWR_MPU_RAM_LOW_SPEED_THRESHOLD)
+			io_setbits32(stm32_pwr_base() + PWR_CR1_OFF,
+				     PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+		if (set_opp_voltage(cpu_opp.current_opp))
+			panic();
+
+		return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result stm32_cpu_opp_set_level(unsigned int level)
+{
+	unsigned int current_level = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int opp = 0;
+
+	assert(cpu_opp.dvfs);
+
+	DMSG("Set OPP to %ukHz", level);
+
+	if (!cpu_opp.dvfs)
+		return TEE_ERROR_GENERIC;
```

Comment:
Either the assertion at line 226 or this test, we don't really need both.
I think we should prefer this test since `stm32_cpu_opp_set_level()` is called straight from the SCMI client interface.